lib/cocoapods/installer/project_cache/target_cache_key.rb

```diff
@@ -0,0 +1,134 @@
+        #
+        attr_reader :hash
+
+        # Initialize a new instance.
+        #
+        # @param [Symbol] type @see #type
+        # @param [Hash{String => Object}] hash @see #hash
+        #
+        def initialize(type, hash)
+          @type = type
+          @hash = hash
+        end
+
+        # Equality function used to compare TargetCacheKey objects to each other.
+        #
+        # @param [TargetCacheKey] other
+        #        Other object to compare itself against.
+        #
+        # @return [Symbol] The difference between this and another TargetCacheKey object.
+        #         # Symbol :none means no difference.
+        #
+        def key_difference(other)
+          if other.type != type
+            :project
+          else
+            case type
+            when :pod_target
+              return :project if (other.hash.keys - hash.keys).any?
+              return :project if other.hash['CHECKSUM'] != hash['CHECKSUM']
+              return :project if other.hash['SPECS'] != hash['SPECS']
+              return :project if other.hash['FILES'] != hash['FILES']
+            end
+
+            this_build_settings = hash['BUILD_SETTINGS_CHECKSUM']
+            other_build_settings = other.hash['BUILD_SETTINGS_CHECKSUM']
+            return :project if this_build_settings != other_build_settings
+
+            this_checkout_options = hash['CHECKOUT_OPTIONS']
+            other_checkout_options = other.hash['CHECKOUT_OPTIONS']
+            return :project if this_checkout_options != other_checkout_options
+
+            :none
+          end
+        end
+
+        def to_h
+          hash
+        end
+
+        # Creates a TargetCacheKey instance from the given hash.
+        #
+        # @param [Hash{String => Object}] hash
+        #        The hash used to construct a TargetCacheKey object.
+        #
+        # @return [TargetCacheKey]
+        #
+        def self.from_cache_hash(hash)
+          cache_hash = hash.dup
+          if files = cache_hash['FILES']
+            cache_hash['FILES'] = files.sort
+          end
+          type = cache_hash['CHECKSUM'] ? :pod_target : :aggregate
+          TargetCacheKey.new(type, cache_hash)
+        end
+
+        # Constructs a TargetCacheKey instance from a PodTarget.
+        #
+        # @param [PodTarget] pod_target
+        #        The pod target used to construct a TargetCacheKey object.
+        #
+        # @param [Bool] is_local_pod
+        #        Used to also include its local files in the cache key.
+        #
+        # @param [Hash] checkout_options
+        #        The checkout options for this pod target.
+        #
+        # @return [TargetCacheKey]
+        #
+        def self.from_pod_target(pod_target, is_local_pod: false, checkout_options: nil)
+          build_settings = {}
+          build_settings[pod_target.label.to_s] = Digest::MD5.hexdigest(pod_target.build_settings.xcconfig.to_s)
```

Comment:
this is from me being burned by this in the past, but could we put the digest algorithm into the cache, so we're free to change it in the future? Also, `MD5` isn't FIPS compliant, if that's something we want to care about

Reply:
Like literally just store a string of the class/method name being used?